File: scala/src/main/scala/fundamentals/Loops.scala
package fundamentals

import scala.annotation.tailrec
import scala.util.control.Breaks
import scala.util.control.Breaks.{break, breakable}

object Loops:
  def demonstrate(): Unit =
    println("=== Loops ===\n")
    
    // While loop
    println("1. While loop:")
    var i = 0
    while i < 5 do
      print(s"$i ")
      i += 1
    println("\n")
    
    // Do-while loop
    println("2. Do-while loop:")
    var j = 0
    do
      print(s"$j ")
      j += 1
    while j < 5
    println("\n")
    
    // For loop with Range
    println("3. For loop with Range:")
    println("a. Inclusive range:")
    for i <- 1 to 5 do
      print(s"$i ")
    println()
    
    println("\nb. Exclusive range:")
    for i <- 1 until 5 do
      print(s"$i ")
    println()
    
    println("\nc. Range with step:")
    for i <- 1 to 10 by 2 do
      print(s"$i ")
    println("\n")
    
    // Nested loops
    println("4. Nested loops:")
    for
      i <- 1 to 3
      j <- 1 to 3
    do
      println(s"($i, $j)")
    println()
    
    // For loop with filters (guards)
    println("5. For loop with filters:")
    for
      i <- 1 to 10
      if i % 2 == 0
      if i > 5
    do
      print(s"$i ")
    println("\n")
    
    // For comprehension (for expressions)
    println("6. For comprehension:")
    val numbers = List(1, 2, 3)
    val letters = List("a", "b", "c")
    
    println("Cartesian product using for:")
    val pairs = for
      n <- numbers
      l <- letters
    yield s"$n$l"
    println(pairs)
    println()
    
    // Breaking loops
    println("7. Breaking loops:")
    println("Using breakable:")
    breakable {
      for i <- 1 to 10 do
        if i == 6 then
          break()
        print(s"$i ")
    }
    println("\n")
    
    // Pattern matching in for loops
    println("8. Pattern matching in for loops:")
    val tuples = List((1, "one"), (2, "two"), (3, "three"))
    for (num, str) <- tuples do
      println(s"Number: $num, String: $str")
    println()
    
    // Recursion example
    println("9. Recursion:")
    def factorial(n: Int): Int =
      if n <= 1 then 1
      else n * factorial(n - 1)
    
    println(s"Factorial of 5: ${factorial(5)}")
    
    // Tail recursion example
    println("\n10. Tail recursion:")
    def factorialTail(n: Int): Int =
      @tailrec
      def loop(x: Int, accumulator: Int): Int =
        if x <= 1 then accumulator
        else loop(x - 1, x * accumulator)
      
      loop(n, 1)
    
    println(s"Factorial of 5 (tail recursive): ${factorialTail(5)}")
    
    // Infinite stream with take
    println("\n11. Working with infinite streams:")
    def fibonacci: LazyList[BigInt] =
      def fib(a: BigInt, b: BigInt): LazyList[BigInt] =
        a #:: fib(b, a + b)
      fib(0, 1)
    
    println("First 10 Fibonacci numbers:")
    fibonacci.take(10).foreach(n => print(s"$n "))
    println("\n")
    
    // foreach loop
    println("12. Foreach loop:")
    val fruits = List("apple", "banana", "orange")
    fruits.foreach(fruit => println(s"I like $fruit"))
    
    println() // Empty line

File: kotlin/src/main/kotlin/fundamentals/Loops.kt
package fundamentals

class Loops {
    fun demonstrate() {
        println("=== Loops ===\n")

        // For loop with range
        println("1. For loop with range:")
        for (i in 1..5) {
            print("$i ")
        }
        println()

        // For loop with downTo
        println("\n2. For loop with downTo:")
        for (i in 5 downTo 1) {
            print("$i ")
        }
        println()

        // For loop with step
        println("\n3. For loop with step:")
        for (i in 0..10 step 2) {
            print("$i ")
        }
        println()

        // For loop with until (exclusive upper bound)
        println("\n4. For loop with until:")
        for (i in 0 until 5) {
            print("$i ")
        }
        println()

        // For loop over array
        println("\n5. For loop over array:")
        val fruits = arrayOf("Apple", "Banana", "Orange", "Mango")
        for (fruit in fruits) {
            print("$fruit ")
        }
        println()

        // For loop with index
        println("\n6. For loop with index:")
        for ((index, fruit) in fruits.withIndex()) {
            println("Index: $index, Value: $fruit")
        }

        // While loop
        println("\n7. While loop:")
        var count = 0
        while (count < 5) {
            print("${count++} ")
        }
        println()

        // Do-while loop
        println("\n8. Do-while loop:")
        count = 0
        do {
            print("${count++} ")
        } while (count < 5)
        println()

        // Break statement
        println("\n9. Break statement:")
        for (i in 1..10) {
            if (i == 6) break
            print("$i ")
        }
        println()

        // Continue statement
        println("\n10. Continue statement:")
        for (i in 1..5) {
            if (i == 3) continue
            print("$i ")
        }
        println()

        // Labeled break
        println("\n11. Labeled break:")
        outer@ for (i in 1..3) {
            for (j in 1..3) {
                if (i == 2 && j == 2) break@outer
                print("[$i,$j] ")
            }
        }
        println()

        // Labeled continue
        println("\n12. Labeled continue:")
        outer@ for (i in 1..3) {
            for (j in 1..3) {
                if (j == 2) continue@outer
                print("[$i,$j] ")
            }
        }
        println()

        // Repeat loop
        println("\n13. Repeat loop:")
        repeat(3) { index ->
            println("Repetition ${index + 1}")
        }

        // forEach loop
        println("\n14. forEach loop:")
        fruits.forEach { fruit ->
            print("$fruit ")
        }
        println()

        // forEachIndexed
        println("\n15. forEachIndexed:")
        fruits.forEachIndexed { index, fruit ->
            println("Index: $index, Value: $fruit")
        }

        // Range iteration with filter and map
        println("\n16. Range with filter and map:")
        (1..10)
            .filter { it % 2 == 0 }
            .map { it * it }
            .forEach { print("$it ") }
        println()

        // Sequence operations
        println("\n17. Sequence operations:")
        (1..5).asSequence()
            .map { it * 2 }
            .filter { it > 5 }
            .forEach { print("$it ") }
        println()

        println() // Empty line
    }
}